Refactor and use MINIMUM_STAKE_DELEGATION constant

brooksprumo opened the pull request:
#### Problem

The minimum stake delegation constant is not used outside of tests.

#### Summary of Changes

* Use the constant in `initialize()`, `delegate()`, `split()`, and `withdraw()`
* Refactor code to be safe and obvious

#### Notes

* No behavior is changed in this PR
* This PR builds on the tests added in #23259 and #23303 to ensure no behavior is changed
* Subsequent tasks for minimum stake delegation are being tracked in #22559
* I recommend _ignoring whitespace_ when looking at the diff here on github

brooksprumo commented:
... or should I name this minimum stake _delegation_?

jstarry commented:
We want to capture a "minimum stake delegation" and that constant needs to be reflected when delegating stake and splitting stake.

brooksprumo commented:
@jstarry Constant has been renamed to `MINIMUM_STAKE_DELEGATION`.

jstarry commented:
> @jstarry Constant has been renamed to `MINIMUM_STAKE_DELEGATION`.

Cool but I don't think it makes sense to use this constant the way you're using it for withdraw. Why not use it for delegations and splitting stake like I suggested?

brooksprumo commented:
> Cool but I don't think it makes sense to use this constant the way you're using it for withdraw. Why not use it for delegations and splitting stake like I suggested?

Ok, will do. I was trying to keep this PR as small as possible, then making further changes in an additional PR. I'll pull those in here.

brooksprumo commented:
@joncinque @jstarry This PR is now ready for a re-review!

brooksprumo commented on runtime/src/bank.rs:
```diff
@@ -214,7 +214,7 @@ impl RentDebits {
 
 type BankStatusCache = StatusCache<Result<()>>;
-#[frozen_abi(digest = "FPLuTUU5MjwsijzDubxY6BvBEkWULhYNUyY6Puqejb4g")]
+#[frozen_abi(digest = "7GSpe6AWtKUnXWzBFAcEZ2QLvhAUfY4VeWBhE8YW6hWL")]
```
Bank hash needed to change due to adding the new Instruction Error.

brooksprumo replied:
No longer applicable.

joncinque commented on programs/stake/src/stake_state.rs:
```diff
@@ -6409,6 +6455,236 @@ mod tests {
+
+            for stake_state in &[
+                StakeState::Initialized(meta),
+                StakeState::Stake(meta, just_stake(starting_stake_delegation)),
```
This case is interesting, because we could have a stake whose delegation is currently less than `MINIMUM_STAKE_DELEGATION`, but also has some non-staked lamports that bring it up above the minimum.  If they try to withdraw those extra lamports, they'll be capped too.

All that to say, this works, well done!

joncinque reviewed (commented):
Great work! This is really close, just a few small comments,

brooksprumo commented on programs/stake/src/stake_state.rs:
```diff
@@ -588,82 +598,54 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
-                        self.data_len()? as u64,
-                        split.data_len()? as u64,
-                    );
+                    let validated_info = validated_split_info.unwrap();
```
Yeah, I was going back and forth on this. Since the impl of `validate_split_info()` guarantees that `unwrap()` is safe here (always `Some` in this variant), I was treating this as a programmer bug and not an instruction error. Maybe an `assert!()` is better here (which is basically what the `unwrap()` is), or a comment? Wdyt?

joncinque replied:
I don't feel that strongly about this, so we can leave it as an `unwrap()` if you think that's better

brooksprumo commented on programs/stake/src/stake_state.rs:
```diff
@@ -820,6 +803,69 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+    dest_rent_exempt_reserve: u64,
+}
+
+fn validate_split_amount(
```
Yah, I feel the same way. I wanted some way to refactor and deduplicate the code that was validating the split amounts, and this was the least-bad thing I thought of.

ryoqun replied:
yeah, these odd rent calc is unfortunate thing from the past (there were no rent sysvar without breaking instruction abi and web3.js had a bug which caused 4000-byte stake accounts and the stake program wrongly accepted that bad accounts...)

brooksprumo commented on programs/stake/src/stake_state.rs:
```diff
@@ -6409,6 +6455,236 @@ mod tests {
+
+            for stake_state in &[
+                StakeState::Initialized(meta),
+                StakeState::Stake(meta, Stake::default()),
```
Yeah... I wasn't sure on this either, since in theory it shouldn't be possible. I'll go ahead and remove it, since `delegate()` already has code to prevent this.

brooksprumo replied:
Done!

brooksprumo commented on programs/stake/src/stake_state.rs:
```diff
@@ -6409,6 +6455,236 @@ mod tests {
+            };
+            // The source account's starting balance is equal to *both* the source and dest
+            // accounts' *final* balance
+            let source_starting_balance =
+                source_stake_delegation + dest_stake_delegation + rent_exempt_reserve * 2;
```
Is this case impossible? If delegations have a minimum of `MINIMUM_STAKE_DELEGATION`, and we want to test a full split, then shouldn't the smallest split possible also be `MINIMUM_STAKE_DELEGATION` (not counting rent exempt reserve)?

Or would this be a test for the future, i.e. if the minimum delegation is raised? So an existing stake account may have a delegation less than the new minimum, and then it should not be possible to split the full amount into a new account?

brooksprumo replied:
For the sake of future-proof-ness, I'm going to add this test.

brooksprumo commented on programs/stake/src/stake_state.rs:
```diff
@@ -820,6 +803,69 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+            let source_minimum_balance = meta
+                .rent_exempt_reserve
+                .saturating_add(MINIMUM_STAKE_DELEGATION);
+            let source_remaining_balance = source_lamports - lamports;
```
Done!

joncinque reviewed (dismissed):
looks good to me! perhaps @CriesofCarrots or @ryoqun could also take a look since this code is pretty crucial, and I believe they did the initial work around here

CriesofCarrots commented on programs/stake/src/stake_state.rs:
```diff
@@ -6431,6 +6482,294 @@ mod tests {
+            (MINIMUM_STAKE_DELEGATION, Ok(())),
+            (
+                MINIMUM_STAKE_DELEGATION - 1,
+                Err(InstructionError::StakeDelegationTooSmall),
```
This makes me nervous, because we actually need to assert that with current code, this error never gets hit (or else we need to feature-gate it).
I believe this can only happen with direct manipulation of the accounts, like you're doing in this test (can you add a comment to that effect here?), and that other tests should cover whether Initialize or Initialize+Withdraw or Initialize+Delegate+Withdraw could leave an account vulnerable to hitting this error. Can you confirm?

brooksprumo replied:
I'll go through and double check that this error cannot be hit without direct manipulation. Additionally, would it be better to just feature-gate this PR then?

CriesofCarrots replied:
I'm a little torn, but the risk of other errors because of cognitive cost of adding the feature gate and then having to work around it or remove it is non-zero. So if we can all convince ourselves that the error can't be hit, let's not :)

brooksprumo replied:
Good call. I've added a comment on this test. Can you let me know what you think?

CriesofCarrots replied:
Very helpful comments, thank you! No more critiques from me

CriesofCarrots commented on programs/stake/src/stake_state.rs:
```diff
@@ -542,8 +544,12 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
             StakeState::Initialized(meta) => {
                 meta.authorized.check(signers, StakeAuthorize::Staker)?;
+                let stake_amount = self.lamports()?.saturating_sub(meta.rent_exempt_reserve); // can't stake the rent ;)
+                if stake_amount < MINIMUM_STAKE_DELEGATION {
+                    return Err(InstructionError::StakeDelegationTooSmall);
+                }
```
Can you please add a comment explaining why this change does not need feature-gating?

brooksprumo replied:
Comment added! Can let me know what you think?

CriesofCarrots commented on programs/stake/src/stake_state.rs:
```diff
@@ -553,9 +559,13 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
             StakeState::Stake(meta, mut stake) => {
                 meta.authorized.check(signers, StakeAuthorize::Staker)?;
+                let stake_amount = self.lamports()?.saturating_sub(meta.rent_exempt_reserve); // can't stake the rent ;)
+                if stake_amount < MINIMUM_STAKE_DELEGATION {
+                    return Err(InstructionError::StakeDelegationTooSmall);
+                }
```
Comment here too, please

brooksprumo replied:
Comment added! Can let me know what you think?

CriesofCarrots reviewed (approved):
I'm requesting a couple comments, but the logic looks correct to my eye. I do think we should get @ryoqun 's eyes+approval also 🙏

t-nelson commented:
Can you update the title and description?  There are clearly logical changes here and I don't want to assess them without understanding the motivation

brooksprumo commented:
@t-nelson 
> Can you update the title and description? There are clearly logical changes here and I don't want to assess them without understanding the motivation

Ok, I've updated the title and description.

In theory there actually are no changes to the current behavior. This PR really is future-proofing and future-prepping changes to the minimum stake delegation. I added this bit to the description:

> Additionally, minimum delegation is not checked beyond `initialize()`, so if the minimum is ever raised there may be inconsistent/unexpected behavior with pre-existing accounts that have a delegation smaller than the new minimum.

Does that feel sufficient? If not I'll happily give it another go to explain it more/more clearly.

brooksprumo commented:
@ryoqun Would you be able to take a look at this PR too? Thanks in advance!

ryoqun commented:
(I'll review this later in full).
 @brooksprumo firstly, i think you need to plumb feature gate? (new error variant should cause differing bank hash under mixed validators?

brooksprumo commented:
> firstly, i think you need to plumb feature gate? (new error variant should cause differing bank hash under mixed validators?

Ah OK. Should I just put the feature gate around the new error variant then? And for the feature-not-enabled branch, would using the InsufficientFunds error be good, or is there a better solution?

ryoqun commented:
> > firstly, i think you need to plumb feature gate? (new error variant should cause differing bank hash under mixed validators?
> 
>  And for the feature-not-enabled branch, would using the InsufficientFunds error be good, or is there a better solution?

yes, if the old branch always results in `InsufficientFunds`. maybe a bit tedious but we need to retain exactly same behavior until the feature activation.

CriesofCarrots commented:
@ryoqun , recommend you take a closer look first. There actually shouldn't be any behavior changes here (although any changes to `MINIMUM_STAKE_DELEGATION` will need to be feature-gated for sure).

ryoqun commented:
> @ryoqun , recommend you take a closer look first. There actually shouldn't be any behavior changes here (although any changes to `MINIMUM_STAKE_DELEGATION` will need to be feature-gated for sure).

oops. seems so. sorry for false alarm...

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -553,9 +568,22 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
             }
             StakeState::Stake(meta, mut stake) => {
                 meta.authorized.check(signers, StakeAuthorize::Staker)?;
+                let stake_amount = self.lamports()?.saturating_sub(meta.rent_exempt_reserve); // can't stake the rent ;)
+                if stake_amount < MINIMUM_STAKE_DELEGATION {
+                    // Since (1) delegation uses the full balance (minus rent_exempt_reserve), and
+                    // (2) stake accounts must be created with a minimum balance of
+                    // rent_exempt_reserve + MINIMUM_STAKE_DELEGATION, then we can be sure that the
+                    // stake_amount will always be at least MINIMUM_STAKE_DELEGATION.
+                    //
+                    // This error case can only be triggered if (1) the MINIMUM_STAKE_DELEGATION is
+                    // raised in the future and this account was created before the raise, or (2)
+                    // during tests where direct manipulation of the account allows us to break
+                    // invariants.
+                    return Err(InstructionError::StakeDelegationTooSmall);
+                }
```
this

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -542,8 +544,21 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
         match self.state()? {
             StakeState::Initialized(meta) => {
                 meta.authorized.check(signers, StakeAuthorize::Staker)?;
+                let stake_amount = self.lamports()?.saturating_sub(meta.rent_exempt_reserve); // can't stake the rent ;)
+                if stake_amount < MINIMUM_STAKE_DELEGATION {
+                    // Since (1) delegation uses the full balance (minus rent_exempt_reserve), and
+                    // (2) stake accounts must be created with a minimum balance of
+                    // rent_exempt_reserve + MINIMUM_STAKE_DELEGATION, then we can be sure that the
+                    // stake_amount will always be at least MINIMUM_STAKE_DELEGATION.
+                    //
+                    // This error case can only be triggered if (1) the MINIMUM_STAKE_DELEGATION is
+                    // raised in the future and this account was created before the raise, or (2)
+                    // during tests where direct manipulation of the account allows us to break
+                    // invariants.
+                    return Err(InstructionError::StakeDelegationTooSmall);
+                }
```
how about factoring out this into `fn validate_stake_amount()` just like `fn valiate_split_amount()` to avoid this somewhat long comment copy-paste here?: https://github.com/solana-labs/solana/pull/22663/files?diff=split&w=1#r802606743

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -553,9 +568,22 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
                 meta.authorized.check(signers, StakeAuthorize::Staker)?;
+                let stake_amount = self.lamports()?.saturating_sub(meta.rent_exempt_reserve); // can't stake the rent ;)
+                if stake_amount < MINIMUM_STAKE_DELEGATION {
+                    // Since (1) delegation uses the full balance (minus rent_exempt_reserve), and
+                    // (2) stake accounts must be created with a minimum balance of
+                    // rent_exempt_reserve + MINIMUM_STAKE_DELEGATION, then we can be sure that the
+                    // stake_amount will always be at least MINIMUM_STAKE_DELEGATION.
```
it seems that there's counter example: https://explorer.solana.com/address/21mQHA6YZg5iX9KAsro4sk76D8jWwov1K64uKLDChV6J

```
$ ./target/release/solana stake-account 21mQHA6YZg5iX9KAsro4sk76D8jWwov1K64uKLDChV6J
Balance: 0.00228288 SOL
Rent Exempt Reserve: 0.00228288 SOL
Delegated Stake: 0 SOL
Active Stake: 0 SOL
Activating Stake: 0 SOL
Stake activates starting from epoch: 181
Delegated Vote Account Address: 9c5bpzVRbfsYY2fannb4hyX5CJUPg3BfH2cL6sR7kJM4
Stake Authority: 488StDvQbiDZeLSAqmpBz2zvjGe1LaAqwZckS21vufBP
Withdraw Authority: 488StDvQbiDZeLSAqmpBz2zvjGe1LaAqwZckS21vufBP
```

ryoqun replied:
as a quick context, there's long-tall tragedy of the ugly current state of code and live data for stake accounts. ;)

ryoqun replied:
i haven't dug in detail. but fully-deactivated withdraw allow this condition to occur?

brooksprumo replied:
Oh... Uh... I'm not sure. @CriesofCarrots or @joncinque do you know?

joncinque replied:
Even from a deactivated stake, you can withdraw up to reserve + 1 lamport. There's currently code to ensure that the reserve is kept, which was correctly updated: https://github.com/solana-labs/solana/blob/86d465c53137978b464e01c1e1155b9f58b166cb/programs/stake/src/stake_state.rs#L799

It was possible to split down to just reserve before, but that was fixed.  This account is probably a holdover from back then, but thankfully it's possible to withdraw all the lamports out of it

brooksprumo replied:
Great! Marking this as resolved. Thanks!

CriesofCarrots replied:
What Jon said :)

I believe the only operations this account will allow would be to withdraw to zero, or add lamports >= 1. @brooksprumo , if this one particular stake account tried to withdraw something less than 0.00228288 SOL, which error would it return?

@ryoqun , any idea how many stake accounts are in this state? Are there any that still exist with rent-exempt reserve of greater than 0.00228288 SOL?

brooksprumo replied:
@CriesofCarrots As long as the account is not `Uninitialized`, then the error will be `InsufficientFunds`
https://github.com/solana-labs/solana/blob/a79819978cb507a8f8f49bbf6e2b37a13704ccea/programs/stake/src/stake_state.rs#L806

If the account _is_ `Uninitialized`, then any amount can with withdrawn (up to the actual balance, of course).

Can these old accounts ever be `Uninitialized`?

CriesofCarrots replied:
Cool, thanks for tracing that for me. Just wanted to check that we couldn't end up with the new error somehow.
Yeah, the weird accounts could be `Uninitialized`. But then their only viable actions would be to keep withdrawing until they reach zero, or deposit until they reach rent-exemt-minimum+1, so no change in behavior.

ryoqun replied:
> Even from a deactivated stake, you can withdraw up to reserve + 1 lamport. There's currently code to ensure that the reserve is kept,

well, i haven't tested locally, but it's possible to create 0-lamports delegated stake account if i read correctly?

https://github.com/solana-labs/solana/blob/a79819978cb507a8f8f49bbf6e2b37a13704ccea/programs/stake/src/stake_state.rs#L754-L768

fully-deactivated stake account should fall into this match arm and return the tuple `(<lockup>, 0+rent_exempt_reserve, false)`

then

https://github.com/solana-labs/solana/blob/a79819978cb507a8f8f49bbf6e2b37a13704ccea/programs/stake/src/stake_state.rs#L802-L806

this `if` shouldn't be triggered  because `withdraw_amount + rent_exempt_reserve > account_balance` => `false` (note: not `>=` here) for all withdrawn lamports but the rent_exempt_reserve


https://github.com/solana-labs/solana/blob/d5dec989b93fc94c9416a60193e52e8b68bd7133/programs/stake/src/stake_state.rs#L554-L568

then, this `stake_lamports` just be `0` without causing errors when redelegating such a stake account later, as this is _saturating_ subtraction.

thus, this can introduce unintended behavior change maybe by adding new `if` in this pr?

ryoqun replied:
@brooksprumo yeah, initialize works as intended. however, stake accounts can be recycled after fully deactivated bypassing the `initialize` altogether like this:

here's more detailed steps of https://github.com/solana-labs/solana/pull/22663#discussion_r802632124

1. create a normal stake account with good funding
2. delegate it to some validator
3. deactivate it
4. withdraw all but `rent_exempt_reserve`
5. delegate it to some validator again

i've just confirmed locally that you can still create 0-stake active stake accounts this way.

ryoqun replied:
@CriesofCarrots 

> @ryoqun , any idea how many stake accounts are in this state?

really small (10 accounts):

```
$ less mainnet-beta-272-2.ledger-tool-inflation-validation-minimum-stake-delegation.stderr1 | grep -E 'account|deacti' | grep -E 'N/A|account' | grep -B1 N/A | grep -v -- --
    account: "12FFUP9zeUTDYvYkv8md2mJHsFD27pUfHFZCW9XTRHba",
    deactivation_epoch: "N/A",
    account: "21mQHA6YZg5iX9KAsro4sk76D8jWwov1K64uKLDChV6J",
    deactivation_epoch: "N/A",
    account: "28AW2otHAuyYXfMrLqLqiKKrA4csMXnFBXEzuzepNC7w",
    deactivation_epoch: "N/A",
    account: "2CSZmTsaTon1TMDSnuJ7vZ3ixM27DgCUDPaEutwosXtg",
    deactivation_epoch: "N/A",
    account: "6V3depRzCh49de2GaxMJRLa9oeZjhe83mrphRaLP42jp",
    deactivation_epoch: "N/A",
    account: "8ZwXhv54DfTWB96PvK3tCyiXRXuGCUbEGBXz9YNrYNud",
    deactivation_epoch: "N/A",
    account: "BqaFpYxiy6MH4Z4wPVh32Tvo6M7kR4dddw9WMyGiu7Fe",
    deactivation_epoch: "N/A",
    account: "Cs9pJKWUPuhBibvbSFizYVGqJDe6enkD2ountgBNi9SN",
    deactivation_epoch: "N/A",
    account: "HrJA7JMzZnf2YmqWQz5N2ke2MpgzCAFtyzrA2yGsAh64",
    deactivation_epoch: "N/A",
    account: "HtTouwfJPWudZj2gRY7Q6RoyfBi25EV5sYV6Q65XZNo4",
    deactivation_epoch: "N/A",
```

however, there's several accounts (~178), which could be 0-lamports delegated stakes at any time (currently residing at the step 4 from https://github.com/solana-labs/solana/pull/22663#discussion_r804319877)

```
$ less mainnet-beta-272-2.ledger-tool-inflation-validation-minimum-stake-delegation.stderr1 | grep -E 'deacti' | grep -v N/A | wc -l
178
```

>  Are there any that still exist with rent-exempt reserve of greater than 0.00228288 SOL?

yeah, there is some (~0.1%).

```
# stake account count by its rent-exempt reserve field for mainnet-beta epoch 272
$ cat mainnet-beta-272-2.csv | grep Stake11111111111111111111111111111111111111 | tr , \  | awk '{print $12}' | sort | uniq -c
 691171 2282880
    673 28786560

```

here's a pointer (https://github.com/solana-labs/solana/pull/13461) for those who don't know the history. :)

now that realloc and sysvars are available, i think it's time to clean up this mess finally? ;) I'm bit reluctant though, considering its benefits.

brooksprumo replied:
@ryoqun Thanks for the detailed issue and steps to reproduce. For this PR, I'm not sure what the correct behavior should be. Here's the fn that validates the stake amount:

https://github.com/solana-labs/solana/blob/a79819978cb507a8f8f49bbf6e2b37a13704ccea/programs/stake/src/stake_state.rs#L826-L846

And here's `delegate()` from master:

https://github.com/solana-labs/solana/blob/c078ca3fb3e203dbd4f3a9556166695e8223cea7/programs/stake/src/stake_state.rs#L530-L569

Should I basically (1) re-write this comment, and (2) replace the `Err` with a `warn!()`?

https://github.com/solana-labs/solana/blob/a79819978cb507a8f8f49bbf6e2b37a13704ccea/programs/stake/src/stake_state.rs#L834-L843

Then I think that would retain the current behavior.

Then I could go the full feature-gate for a subsequent PR that puts the error back. Would that work?

t-nelson replied:
If in doubt, just gate it.  No sense bending over backwards avoiding it just to fork the cluster on a corner case anyway

ryoqun replied:
> Then I could go the full feature-gate for a subsequent PR that puts the error back.

yeah. sounds good. the additional bonus would be add tests in this pr to validate the current behavior for the newly-found minimum stake delegation loophole (i.e. withdraw & re-delegate with fully-deactivated & only-just-enough-rent-exempt-reserve) so that the subsequent feature gate pr's behavior change should be validated just by tweaking the `assert!(..)`s in the tests.

> Should I basically (1) re-write this comment

Yeah.

> and (2) replace the `Err` with a `warn!()`?

Well, how much `warn!()` would be beneficial to node operators? i think this just scares them for no good reason because those `warn!()` are just not actionable for them. also, this will allow little annoyance of allowing log spamming by attackers.

brooksprumo replied:
Done!

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -542,8 +544,21 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+                    // stake_amount will always be at least MINIMUM_STAKE_DELEGATION.
+                    //
+                    // This error case can only be triggered if (1) the MINIMUM_STAKE_DELEGATION is
+                    // raised in the future and this account was created before the raise, or (2)
```
also, slashing might be worth to mention?

brooksprumo replied:
Yes, good point. Will do.

brooksprumo replied:
Done!

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -610,82 +638,59 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
-                    );
+                    // SAFETY: `.unwrap()` is safe here, since `validate_split_amount()` always
+                    // returns `Some` when the state is `StakeState::State`.
+                    let validated_split_info = validated_split_info.unwrap();
```
this

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -610,82 +638,59 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
         }
 
         if let StakeState::Uninitialized = split.state()? {
-            // verify enough account lamports
-            if lamports > self.lamports()? {
-                return Err(InstructionError::InsufficientFunds);
-            }
```
nits: how about restoring this not-duplicate check to avoid the scary `.unwrap()` (https://github.com/solana-labs/solana/pull/22663/files?diff=split&w=1#r802645422 and https://github.com/solana-labs/solana/pull/22663/files?diff=split&w=1#r796146060)

1. call `validate_split_amount()` in this `fn` from both `match` arms for `StakeState::Stake` and `StakeState::Initialize` respectively.
2. for that adjust the signature like this `fn validate_split_amount(self, split, lamports, meta)`.
3. in turn, we can remove the `match` in the `validate_split_amount` to avoid `.unwrap()` there as well.
4. profit with type safety. :)

brooksprumo replied:
Love it. Will do.

brooksprumo replied:
Done!

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -842,6 +848,69 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+                // nothing to do here
+            } else if source_remaining_balance < source_minimum_balance {
+                // the remaining balance is too low to do the split
+                return Err(InstructionError::InsufficientFunds);
```
(i wonder we should start to use StakeDelegationTooSmall or not for consistent behavior...)

brooksprumo replied:
I thought about this too. There are multiple error cases:

1. Split amount leaves src account below rent exempt minimum
2. Split amount leaves dst account below rent exempt minimum
3. Split amount leaves src account below (rent exempt minimum + minimum stake delegation)
4. Split amount leaves dst account below (rent exempt minimum + minimum stake delegation)

What should each error be?

(1) and (2) I think are straight forward to leave as erroring with InsufficientFunds. (3) and (4) are more up in the air. I ended up settling on the idea that the split amount _does_ have insufficient funds to satisfy both rent exempt minimum _and_ the minimum stake delegation, which is how the code works today. I agree it may be beneficial to have further context returned in the error to know if the insufficient funds were due to not meeting rent exempt minimum, or if it was for not meeting rent exempt minimum + minimum stake delegation.

Since this PR is not changing current behavior, maybe this can turn into its own separate issue and PR to resolve it. Wdyt about separate issue? Also, wdyt about which errors to use?

ryoqun replied:
wow quite good writeup for summarizing the current state regarding this. thanks

so, next question is why did we introduce `StakeDelegationTooSmall` in this pr in the first place? can't we just reuse (i.e. = overload...) the `InsufficientFunds`?

don't take me wrong. i love context-sensitive detailed errors. so, i like `StakeDelegationTooSmall` than the too generic `InsufficientFunds`. however, avoiding consistency is more important, in my opinion. this benefits to developers for learning and ensuring security with consistent reasoning

conceptually, trying to creat a new stake account should functionally be equivalent to splitting full amount from a existing stake account. so ideally same error should be returned for both cases if lamports are too small.

currently, both code path results in `InsufficientFunds`. however, starting from this pr, the creating case starts to return `StakeDelegationTooSmall`, right? I wonder we rather want to stick to `InsufficientFunds` for now. and later address the context-sensitive errors if wanted as a separate issue?

brooksprumo replied:
Yes, I can do that. I'll get working on that ASAP and also open a new issue to track adding the StakeTooSmall error.

brooksprumo replied:
Added a task in https://github.com/solana-labs/solana/issues/22559 to track adding the StakeDelegationTooSmall error.

ryoqun commented:
> @ryoqun Ok, I've hopefully addressed all your comments (thank you them!); can you take another look?

yeah, the cleanup commits are really good ones :100: i like the newer code very much. :)

brooksprumo commented on programs/stake/src/stake_state.rs:
```diff
@@ -842,6 +817,83 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+    stake_amount: u64,
+}
+
+/// Ensure the stake delegation amount is valid.  This checks that the account meets the minimum
+/// balance requirements to delegate stake.  If not, return an error.
+fn validate_delegate_amount(
+    account: &KeyedAccount,
+    meta: &Meta,
+) -> Result<ValidatedDelegateInfo, InstructionError> {
+    let stake_amount = account.lamports()?.saturating_sub(meta.rent_exempt_reserve); // can't stake the rent ;)
+    Ok(ValidatedDelegateInfo { stake_amount })
+}
```
In a subsequent PR, this function will be expanded to enforce that the stake amount is at least the minimum stake delegation. But for now, to not change the current behavior, skip that check.

brooksprumo commented:
@ryoqun @CriesofCarrots @joncinque this PR is ready for another round of reviews. I've done a few updates:
1. I've implemented and merged the tests in PR #23259 separately
2. This PR now does not change any behavior from the status quo
3. Subsequent tasks for changing behavior are being tracked in #22559

So this PR is basically now just a refactor (and should be much simpler).

CriesofCarrots commented on programs/stake/src/stake_state.rs:
```diff
@@ -842,6 +817,83 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+}
+
+/// Ensure the stake delegation amount is valid.  This checks that the account meets the minimum
+/// balance requirements to delegate stake.  If not, return an error.
+fn validate_delegate_amount(
```
```suggestion
/// balance requirements of delegated stake.  If not, return an error.
fn validate_delegated_amount(
```
I know you're using "delegate" as a verb, but I found this confusing due to ambiguity with "delegate" the noun, which means something different.

brooksprumo replied:
Works for me! I went back and forth on this myself, so, happy to go along with whatever version you prefer.

CriesofCarrots commented on programs/stake/src/stake_state.rs:
```diff
@@ -842,6 +817,83 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
     }
 }
 
+/// After calling `validate_delegate_amount()`, this struct contains calculated values that are used
+/// by the caller.
+struct ValidatedDelegateInfo {
```
```suggestion
/// After calling `validate_delegated_amount()`, this struct contains calculated values that are used
/// by the caller.
struct ValidatedDelegatedInfo {
```

CriesofCarrots replied:
Remind why we need this to be a struct instead of just returning u64? More info coming?
Seems like we parse it to pull out the stake_amount immediately in all cases.

brooksprumo replied:
Yeah, this is another one of those fuzzy areas...

The main role of the function is to validate the stake delegation amount. When I see a fn called `validate_delegated_amount()`, it's not immediately obvious (to me) what a `u64` return value would represent. As an specific example, take the function `fn add(x: u64, y: u64) -> u64`, it is immediately clear what this function (likely) returns.

So in the case of `validate_delegated_amount()`, that function calculates what the stake amount is as part of its validation. I forget where I read it, but someone said (paraphrasing):
> Return all values calculated within a function.
Returning _all_ values is likely an exaggeration (or at least impractical for performance reasons), but returning _useful_ calculated values does make sense, since I need the stake amount after validation and I don't want to repeat the calculation again. I like to put these calculated values into an "info" struct for two reasons:

1. The struct communicates the intent; that its fields are informational/side-band, and computed during fn execution.
2. A struct and its fields have names, so it becomes self-documenting. I don't need to flip back to the function's implementation to know what the value represents, as the field's name tells me so. (Assuming the names are descriptive enough (or if there are doc comments added as well).)
 
This is actually something I've found challenging when reading through new areas of the code base, when functions have return values like `(u64, u64, u64, Vec<(Pubkey, u64)>)` and I try to trace that through the stack.

Since this function is not likely a performance bottleneck, I would not expect a u64 wrapped in a struct to result in performance degradation (in fact the Release build should produce identical ASM ([godbolt link](https://godbolt.org/z/TvGhThMcP)).) 

I guess the tl;dr is that there's no performance impact, and it improves (my) comprehension and intent.

t-nelson replied:
> when functions have return values like `(u64, u64, u64, Vec<(Pubkey, u64)>)`

All tuples of size 3+ need to be eradicated from the tree in favor of `structs`

CriesofCarrots replied:
I'm with you about structifying tuples generally, but I personally think this struct provides more obfuscation than elucidation. 🤷‍♀️ Not a hill I'll die on; take it or leave it ;)

brooksprumo replied:
Oh! I forgot one. I wrote `validate_split_amount()` first, which has two fields in its struct. So when I wrote `validate_delegated_amount()`, I decided to use the same pattern since it was doing something similar, even though there was only one field in the struct. (FWIW)

CriesofCarrots commented on programs/stake/src/stake_state.rs:
```diff
@@ -842,6 +817,83 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+    account: &KeyedAccount,
+    meta: &Meta,
+) -> Result<ValidatedDelegateInfo, InstructionError> {
+    let stake_amount = account.lamports()?.saturating_sub(meta.rent_exempt_reserve); // can't stake the rent ;)
```
```suggestion
    let stake_amount = account.lamports()?.saturating_sub(meta.rent_exempt_reserve); // can't stake the rent
```
Not sure that winky face is communicating anything useful anymore

CriesofCarrots commented on programs/stake/src/stake_state.rs:
```diff
@@ -842,6 +817,83 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+/// by the caller.
+#[derive(Copy, Clone, Debug, Default)]
+struct ValidatedSplitInfo {
+    src_remaining_balance: u64,
+    dst_rent_exempt_reserve: u64,
```
Please spell out things everywhere

brooksprumo replied:
Done!

CriesofCarrots commented on programs/stake/src/stake_state.rs:
```diff
@@ -842,6 +817,83 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+/// delegation, and that the source account has enough lamports for the request split amount.  If
+/// not, return an error.
+fn validate_split_amount(
+    src_account: &KeyedAccount,
+    dst_account: &KeyedAccount,
```
Please spell out. Can probably drop _account suffix, though, if you're worried about length

brooksprumo replied:
Done!

CriesofCarrots commented on programs/stake/src/stake_state.rs:
```diff
@@ -842,6 +817,83 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+        return Err(InstructionError::InsufficientFunds);
+    }
+
+    // verify that the source account still has enough lamports left
```
```suggestion
    // verify that the source account still has enough lamports left:
```

CriesofCarrots reviewed (approved):
Refactor looks good to me! Just some opinions about variable and method names from me

joncinque commented on programs/stake/src/stake_state.rs:
```diff
@@ -842,6 +821,84 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+    );
+    let destination_minimum_balance =
+        destination_rent_exempt_reserve.saturating_add(MINIMUM_STAKE_DELEGATION);
+    if lamports < destination_minimum_balance {
+        return Err(InstructionError::InsufficientFunds);
+    }
```
Please correct me if I'm wrong, but I think this could incorrectly fail if the destination has some lamports already.  For example, I can `transfer` the rent exempt reserve into `split`, and then separately `split` `MINIMUM_STAKE_DELEGATION` into it, and that would fail

CriesofCarrots replied:
Oh that's a great catch, Jon. We seem to be missing `.saturating_sub(split.lamports()?)` from the old L628 and L680 in the destination_minimum_balance calculation
Sorry for missing that @brooksprumo !

brooksprumo replied:
Will fix. Phew! Glad y'all caught this. I'll also add a test for this.

CriesofCarrots replied:
I'm also a little surprised that existing tests didn't catch this, as we used to have a bunch of cases for destination accounts with existing lamport balances. I'll take a peek.

joncinque replied:
Yeah this really surprised me too, which is why I wasn't sure about my comment :sweat_smile:

CriesofCarrots replied:
😞 looks like just a gap in the testing. All our tests with various destination starting-balance cases are for splitting too much or just enough. @brooksprumo , how about I pr those missing test cases (probably an extension of [this test](https://github.com/solana-labs/solana/blob/2e3eafaa116fe2f217767af4412463103c77ae3c/programs/stake/src/stake_state.rs#L4049)), and then you can rebase this PR to sanity check?

brooksprumo replied:
This issue has now been resolved in this PR. Tests have been added to PR #23303. Once that PR is merged, I'll rebase this PR to confirm no behavior breakages.

joncinque reviewed (commented):
Just the one comment, looks good otherwise!

brooksprumo commented:
@CriesofCarrots @joncinque @ryoqun This PR is ready for another review. Hopefully I've ironed out all the kinks this time! Thanks again for all the help and time on this!

CriesofCarrots reviewed (dismissed):
I didn't spot anything else; thanks for taking such a careful approach to this!

brooksprumo commented:
@joncinque @ryoqun Let me know if you'd like to review again before I merge—I'm happy to wait! Wanted to check in and make sure, and _not_ make assumptions. Otherwise I'll plan to merge in ~a day.

ryoqun commented:
> @joncinque @ryoqun Let me know if you'd like to review again before I merge—I'm happy to wait! Wanted to check in and make sure, and _not_ make assumptions. Otherwise I'll plan to merge in ~a day.

@brooksprumo hi, I'm planning to finish to review this in its entirety in a day. this pr looks really prime for the merge so far. sorry for late review...

brooksprumo commented:
> sorry for late review...

@ryoqun No worries, not late at all. Thanks for all the time you've put into this!

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -610,84 +616,56 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
 
         if let StakeState::Uninitialized = split.state()? {
-            // verify enough account lamports
-            if lamports > self.lamports()? {
```
hmm, if i read the code correctly, i think this `if` shouldn't be removed as i concerned here a bit (https://github.com/solana-labs/solana/pull/22663/files#r802648796).

Removing this `if` condition will introduce unintended behavior change regarding `StakeState::Uninitialized`.

specifically, newer code will wrongly allow multiple invocations of `StakeInstruction::Split`s on **0-lamport** already `Uninitialized` stake accounts as `self` here in a single tx. That behavior itself is a meaningless and no-op. However, older code will abort at the 2nd `Split` instruction immediately. Thus resulting in tx status outcome change.

That's because the subsequent codepath for `StakeState::Uninitialized` doesn't prevent such an nonsense action and just run as no-op, after removing this (if i read the code).

Maybe, we should restore this check here because it's rather general check like the preceeding `.owner()` and `.data_len()`. it's easy to miss all of following `match` arms are properly guarded as before like when adding newer `StakeState` variant in the future (quite unlikely though).

Could you double-check this? and add test to demonstrate this unintended behavior change?

brooksprumo replied:
@ryoqun Can you help me with the test procedure? Right now I have basically this:

```
- source account set to `Uninitialized` with zero lamports
- dest1 account set to `Uninitialized` with zero lamports
- split source into dest1 with amount = 0; assert Ok
- dest2 account set to `Uninitialized` with zero lamports
- split source into dest2 with amount = 0; assert Err
```

And that returns Ok for both asserts, for both my PR and against master. So I'm guessing I must be missing something.

<details>
<summary>Here's the actual code</summary>

```rust
    #[test]
    fn test_multiple_splits_from_uninitialized() {
        let source_pubkey = Pubkey::new_unique();
        let source_account = AccountSharedData::new_ref_data_with_space(
            0,
            &StakeState::Uninitialized,
            std::mem::size_of::<StakeState>(),
            &id(),
        )
        .unwrap();
        let source_keyed_account = KeyedAccount::new(&source_pubkey, true, &source_account);

        for i in 0..10 {
            let dest_pubkey = Pubkey::new_unique();
            let dest_account = AccountSharedData::new_ref_data_with_space(
                0,
                &StakeState::Uninitialized,
                std::mem::size_of::<StakeState>(),
                &id(),
            )
            .unwrap();
            let dest_keyed_account = KeyedAccount::new(&dest_pubkey, true, &dest_account);

            let result =
                source_keyed_account.split(0, &dest_keyed_account, &HashSet::from([source_pubkey]));
            assert!(result.is_ok());
        }
    }
```

And the test passes for both master and this PR, with the 10 splits.
</details>

jstarry replied:
My read is that the only way to get success with `lamports > self.lamports()?` is if `self.key == split.key` because otherwise `self.try_account_ref_mut()?.checked_sub_lamports(lamports)?` would fail due to underflow. It looks like the change in this PR allows such a transaction to succeed.

ryoqun replied:
gah, as @jstarry said, i lacked the important bit to decipher my previous comment... yeah, i definitely should had mentioned about `self-split`... sorry about confusion.

brooksprumo replied:
Oh, duh, hah I see now. I forgot that `validate_split_amount()` was moved into the match arms... I've re-added the check _above_ the match.

For the test, I still always get `split()` to return an error if `lamports > self.lamports()`, specifically because of the `.checked_sub_lamports()` returning an Underflow error. And calling `split()` with zero lamports always succeeds (on master)...
I feel like I'm still not understanding something properly though.

ryoqun replied:
(hmm, i'll check this shortly after).

ryoqun replied:
@brooksprumo my bad. my above comment had bugs (specifically, the 2nd ix and 0-lamport requirement were total false information). maybe i should just write down the code instead of english.. ;)

here's the updated test, which should demonstrate the differing behavior: 
https://github.com/ryoqun/solana/tree/minimum-stake-constant-test-tweaks

also, your latest pushed commits looks okay.

brooksprumo replied:
@ryoqun Thanks for the code! I should've replied back here last night; @jstarry pointed out my issue, needing to use the same keyed account for both source and destination of the split.

ryoqun replied:
@brooksprumo no problem. could you add this new test into pr for future prosperity? i think we just lacked another edge case here, resulting in unnoticed behavior change.

brooksprumo replied:
@ryoqun Yep on it!

brooksprumo replied:
@ryoqun Tests are over here: https://github.com/solana-labs/solana/pull/23442

brooksprumo replied:
I've also run this branch against those tests. So this PR should be safe to review again. I can ensure the other PR is merged first, and this PR gets rebased and re-runs CI as well.

brooksprumo replied:
Done! Tests have been merged and this PR has been rebased.

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -610,84 +616,56 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+                        validated_split_info.destination_rent_exempt_reserve;
                     split.set_state(&StakeState::Initialized(split_meta))?;
                 }
                 StakeState::Uninitialized => {
```
here

joncinque commented on programs/stake/src/stake_state.rs:
```diff
@@ -842,6 +823,94 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+    );
+    let destination_minimum_balance =
+        destination_rent_exempt_reserve.saturating_add(MINIMUM_STAKE_DELEGATION);
+    let minimum_split_amount =
+        destination_minimum_balance.saturating_sub(destination_account.lamports()?);
```
You can do this in a separate PR, but this could create a strange situation.  Imagine that I have an active stake with a lot of SOL, and I want to split some of it into another account.  If the destination account already has `MINIMUM_STAKE_DELEGATION + destination_rent_exempt_reserve - 1`, I could split 1 active lamport into it, which means that it would have a smaller delegation than `MINIMUM_STAKE_DELEGATION`.

So you might need to have `minimum_split_amount` resolve to `max(MINIMUM_STAKE_DELEGATION, destination_minimum_balance.saturating_sub(destination_account.lamports()?))` if the source account is active

CriesofCarrots replied:
Yeah, that's true. This gets elided at present, since `MINIMUM_STAKE_DELEGATION == 1`. Great spot, Jon

brooksprumo replied:
Ah yes! To my happiness I'm glad that I haven't inadvertently changed any existing behavior! To resolve this, I've added/updated tests over here: https://github.com/solana-labs/solana/pull/23456.

This way, if the minimum stake delegation is ever raised, the tests will fail until this is resolved. Tracking that in #23458.

brooksprumo replied:
Done! Tests were merged, this PR has rebased. And since #23458 has been deleted, I've implemented your fix in this PR as well.

joncinque reviewed (commented):
Looks good to me, just one comment which might be important for enforcing `MINIMUM_STAKE_DELEGATION` on active source stakes

brooksprumo commented:
@ryoqun @joncinque Tests have been added and merged, and this PR rebased. Good for another round of reviews. Thanks in advance!

joncinque commented on programs/stake/src/stake_state.rs:
```diff
@@ -842,6 +823,99 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+        destination_rent_exempt_reserve.saturating_add(MINIMUM_STAKE_DELEGATION);
+    let destination_balance_deficit =
+        destination_minimum_balance.saturating_sub(destination_account.lamports()?);
+    let minimum_split_amount = std::cmp::max(MINIMUM_STAKE_DELEGATION, destination_balance_deficit);
```
Sorry, I don't think I made the contra to this part clear.  There could be a different strange behavior with this, where anytime I split an `Initialized` stake into an account that already has `destination_minimum_balance`, I can't split anything less than `MINIMUM_STAKE_DELEGATION`, which is fine for now, but once the number is big, it means I'll have to do a `withdraw` + `transfer`, which is probably fine, but something to keep in mind.

The best of all behaviors would be to take the `max` only if `source_account` is a `Stake`, which can be expressed through an additional `is_stake` flag or something.  Does that make sense?

brooksprumo replied:
Ah, OK. I glossed over that in your other comment, sorry! I've just pushed another commit that hopefully fixes this.

I put the changes inside the match within `split()`, since I wanted to try and keep `validate_split_amount()` to be state-agnostic. I also did _not_ add a check if splitting 100%, since I'm making the assumption that the source account meets the minimum requirements.

(But note, then what happens if the MINIMUM_STAKE_DELEGATION is raised and there's an existing stake account that tries to split 100%. Should that succeed? Should that fail? I'm not sure. Luckily that doesn't need to be answered immediately for this PR.)

joncinque replied:
> (But note, then what happens if the MINIMUM_STAKE_DELEGATION is raised and there's an existing stake account that tries to split 100%. Should that succeed? Should that fail? I'm not sure. Luckily that doesn't need to be answered immediately for this PR.)

Hehe indeed not for this PR, but my thinking was that only merge / deactivate / withdraw would be supported for existing stakes underneath `MINIMUM_STAKE_DELEGATION`.  You could even talk me out of "merge"

joncinque reviewed (commented):
I'm really sorry that this keeps getting dragged out, but it's really really close!

joncinque reviewed (dismissed):
Looks good to me, thanks for your patience!

brooksprumo commented:
@ryoqun I will wait for your review before merging, unless you say otherwise. (No rush, just wanted to make sure I was explicit!)

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -655,62 +647,59 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
                         // Otherwise, the new split stake should reflect the entire split
-                        // requested, less any lamports needed to cover the split_rent_exempt_reserve
+                        // requested, less any lamports needed to cover the split_rent_exempt_reserve.
+                        let split_lamports = split.lamports()?;
+                        let split_rent_exempt_reserve_deficit = validated_split_info
+                            .destination_rent_exempt_reserve
+                            .saturating_sub(split_lamports);
+                        let minimum_split_amount = MINIMUM_STAKE_DELEGATION
+                            .saturating_add(split_rent_exempt_reserve_deficit);
+                        if lamports < minimum_split_amount {
+                            return Err(InstructionError::InsufficientFunds);
+                        }
```
well, i think this logic is desired to be encapsulated inside `.validate_split_amount()` while being enclosed inside the new `if source_remaining_balance != 0 {...}`.

That's because this is actually another new validation code, which we're trying to refactor all of 'em. :)

ryoqun replied:
I was forgetting that jon is suggesting a similar: https://github.com/solana-labs/solana/pull/22663/files#r818061753 and that the replaced logic is shared for `StakeState::Stake` and `StakeState::Initialized(meta)`....

I'd suggest passing a flag like `is_stake_active` to `validate_split_amount()`...

brooksprumo replied:
Ok, done!

Instead of passing a bool to `validate_split_amount()`, I've instead gone with `Option<&Stake>`. It functions similar to a bool, but now there's additional type safety to ensure there's actually a staked account being split.

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -608,39 +614,25 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+                // lamports minus rent_exempt_reserve if it starts out with a zero balance
+                let (remaining_stake_delta, split_stake_amount) =
+                    if validated_split_info.source_remaining_balance == 0 {
                         // If split amount equals the full source stake, the new split stake must
```
nit: this comment could be tweaked to avoid being stale a bit: `If split amount equals the full source stake as implied by 0 source_remaining_balance, the new split stake ...`.

brooksprumo replied:
Done!

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -655,62 +647,59 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
                             lamports,
-                            lamports - split_rent_exempt_reserve.saturating_sub(split.lamports()?),
+                            lamports.saturating_sub(
+                                validated_split_info
+                                    .destination_rent_exempt_reserve
+                                    .saturating_sub(split_lamports),
```
never mind, considering my comment https://github.com/solana-labs/solana/pull/22663/files#r819635988, maybe fine as-is.

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -842,6 +826,110 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+        return Err(InstructionError::InsufficientFunds);
+    }
+
+    // If the source account is already staked, the destination will end up staked as well.  Verify
+    // the destination account's delegation amount is at least MINIMUM_STAKE_DELEGATION.
+    //
+    // The *delegation* requirements are different than the *balance* requirements.  If the
+    // destination account is prefunded with a balance of `rent exempt reserve + minimum stake
+    // delegation - 1`, the minimum split amount to satisfy the *balance* requirements is 1
+    // lamport.  And since *only* the split amount is immediately staked in the destination
+    // account, the split amount must be at least the minimum stake delegation.  So if the minimum
+    // stake delegation was 10 lamports, then a split amount of 1 lamport would not meet the
+    // *delegation* requirements.
```
i love this crisp and lucid explanation. :)

ryoqun reviewed (approved):
@brooksprumo lgtm
thanks for your patience with my perfectionism and stubbornness. i think this pr is in flawless state now. excellent work.

by the way, what do you think about shipping this? I'd suggest 2 more things to solidify our confidence of those changes:

1. run locally built solana-validator for several epochs against mainnet-beta/testnet/devnet 
2. request neodyme and others to informally review this (now that this is mergeable anytime, it's applicable to bounties here https://github.com/solana-labs/solana/blob/master/SECURITY.md; no need to risk our mainnet-beta; see https://github.com/solana-labs/solana/pull/22546#issuecomment-1035964745 for reference and i ping you on discord as well)

bennofs commented on programs/stake/src/stake_state.rs:
```diff
@@ -842,6 +826,110 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+    account: &KeyedAccount,
+    meta: &Meta,
+) -> Result<ValidatedDelegatedInfo, InstructionError> {
+    let stake_amount = account.lamports()?.saturating_sub(meta.rent_exempt_reserve); // can't stake the rent
```
The function is called `validate_delegated_amount` but uses `saturating_sub` which won't throw an error if the balance is less than rent exempt reserve. Is it possible to use `checked_sub` instead of the saturating one and throw an error when the sub fails, or are there already stake accounts on chain which would fail in that case?

brooksprumo replied:
The current behavior is intentional. (1) In the case where an account's balance is below the rent exempt minimum, then it has zero delegate-able lamports. (2) Adding new errors and/or changing what conditions return errors will change the program's behavior, which would break consensus without a feature-gate. 

(In an initial iteration of this PR I had checks within this function to error in these cases, but I learned that wouldn't be a good idea to add—for the above reasons.)

bennofs replied:
OK, makes sense

bennofs commented on programs/stake/src/stake_state.rs:
```diff
@@ -842,6 +826,110 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+    } else {
+        // all clear!
+        // nothing to do here
+    }
```
this is very picky but you could write this as:

```rust
let source_enough_lamports = source_remaining_lamports >= source_minimum_balance; || source_remaining_balance == 0
if !source_enough_lamports {
  return Err(InstructionError::InsufficientFunds);
}
```

which to me more directly expresses the condition stated in the comment

brooksprumo replied:
Noted. I think I'm going to leave the code as-is so that I don't have to run (and babysit) CI again. Let me know if that's not OK.

bennofs replied:
Yeah that's totally okay

bennofs reviewed (approved):
Looks good to me, just two very picky suggestions, which are more of a personal opinion so feel free to ignore if you don't agree :)

brooksprumo commented:
@ryoqun 

> by the way, what do you think about shipping this? I'd suggest 2 more things to solidify our confidence of those changes:
> 
> 1. run locally built solana-validator for several epochs against mainnet-beta/testnet/devnet (should see no bank hash mismatch then)

Sounds like a good idea to me; I'll get on this. To be clear, I should run validators on all three networks?

> 2. request neodyme and others to informally review this

Also sounds like a good idea!

ryoqun commented:
> @ryoqun
> 
> > by the way, what do you think about shipping this? I'd suggest 2 more things to solidify our confidence of those changes:
> > 
> > 1. run locally built solana-validator for several epochs against mainnet-beta/testnet/devnet (should see no bank hash mismatch then)
> 
> Sounds like a good idea to me; I'll get on this. To be clear, I should run validators on all three networks?

I'd recommend all of them. i know that sounds tedious but the tediousness is 1000x better than the mess of bug at mainent-beta, imo. also, you can spawn 3 gcp instances at the same time and just run your binary there and just wait. I assume this isn't too much of hassle. Previously, I found a bug of stake program only exhibited at testnet's live dataset. (devnet and testnet usually have very odd data; nice exercise for your hot-pressed code). Also this pr's nature is _negative_ feature activation. i mean, its somewhat large code changes must not deviate from the current behavior in every corner of cases. so, it needs dacent track record before reaching to mainnet-beta regardless of urgency (assuming you think this worth backport for upcoming actual change).

on the other hand, merging and v1.10-backporting can comes first to avoid merge conflicts or if it's convenient for you (only pend v1.9-backporting), so that the above general principle is honored. :)

> > 2. request neodyme and others to informally review this
> 
> Also sounds like a good idea!

:+1: seems we got it already: ref: https://github.com/solana-labs/solana/pull/22663#pullrequestreview-900943605 (@bennofs thx!)

brooksprumo commented:
Ok, I've let this PR run on mainnet-beta, testnet, and devnet for multiple epochs. I'm going to merge this in!

* mainnet-beta ran for ~6 days until running into [this unrelated consensus issue](https://github.com/solana-labs/solana/issues/23682)
* testnet ran for 218 hours (just over 9 days) without issue
* devnet ran for 164 hours (just under 7 days) without issue

ryoqun commented:
> Ok, I've let this PR run on mainnet-beta, testnet, and devnet for multiple epochs. I'm going to merge this in!
> 
>     * mainnet-beta ran for ~6 days until running into [this unrelated consensus issue](https://github.com/solana-labs/solana/issues/23682)
> 
>     * testnet ran for 218 hours (just over 9 days) without issue
> 
>     * devnet ran for 164 hours (just under 7 days) without issue

congrats! :) it's really my joy to see this pr to be merged finally. :)
